### PR TITLE
C++: Expose more information in FormattingFunction and make subclasses private.

### PR DESCRIPTION
**Pull request by geoffw0**
Follow-up from https://github.com/github/codeql/pull/4634
 - expose information about how the formatted string is output though new predicates in `FormattingFunction`
 - update uses of subclasses `Printf`, `Fprintf` etc to use the new predicates instead
 - make the subclasses `Printf`, `Fprintf` etc private

The tricky part was getting `UserDefinedFormattingFunction` right (i.e. handling user wrappers to formatting functions).  It was already implemented with mutually recursive predicates `primitiveVariadicFormatter` etc underneath, I believe I've managed to upgrade them and simplify the interface to them without adding much bulk to the implementation.

In the case of `AttributeFormattingFunction` , we simply do not know what is done with the formatted string, so the new predicates are not implemented.

The test is for a special case, where a user wraps `vsprintf` or `vfprintf` in such a way as it doesn't use an output parameter from the caller, and is thus behaving like `vprintf`.  My first attempt to implement the `UserDefinedFormattingFunction` changes was confused by this case.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll [hunk 1]
```diff
@@ -109,9 +109,24 @@ abstract class FormattingFunction extends ArrayFunction, TaintFunction {
...
+   * Holds if this function outputs to a global stream such as standard output,
+   * standard error or a system log. For example `printf`.
+   */
+  predicate isOutputGlobal() { none() }
+  
+  /**
+   * Holds if this function outputs to the stream indicated by
+   * `getOutputParameterIndex()`, that is, this function behaves like `fprintf`.
+   */
+  predicate isOutputStream() { none() }
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll:129]: Could we just implement this as a `final` predicate that checks whether `getOutputParameterIndex()` has a result?

File: cpp/ql/src/semmle/code/cpp/commons/Printf.qll [hunk 2]
```diff
@@ -34,66 +34,77 @@ class AttributeFormattingFunction extends FormattingFunction {
...
-predicate variadicFormatter(Function f, int formatParamIndex) {
-  primitiveVariadicFormatter(f, formatParamIndex)
-  or
-  not f.isVarargs() and
-  callsVariadicFormatter(f, formatParamIndex)
+deprecated predicate primitiveVariadicFormatter(TopLevelFunction f, int formatParamIndex) {
+  primitiveVariadicFormatter(f, _, formatParamIndex, _)
+}
+
+deprecated predicate variadicFormatter(Function f, int formatParamIndex) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/commons/Printf.qll:106]: We require QLDoc on public predicates, even if they're deprecated. See https://github.com/github/codeql/blob/main/docs/qldoc-style-guide.md#deprecated-predicates

**Review by jbj — APPROVED** [review]
Otherwise LGTM

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll:129]: No, this predicate only holds when `getOutputParameterIndex()` points to a stream parameter, not when it points to a buffer parameter.  I'm going to think about a better interface for this, or failing that at least better QLDoc...

File: cpp/ql/src/semmle/code/cpp/commons/Printf.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/commons/Printf.qll]: Ah, good point.  Fixed.

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll]: I've just pushed a commit combining `getOutputParameterIndex` and `isOutputStream` into one `getOutputParameterIndex(boolean isStream)`. I believe this makes it easier to understand what's going on and harder to miss the existence of `isOutputStream` by accident.

Thank you for the question that led me down this road.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll]: I like this interface a lot better. Do we need to keep a deprecated parameter-less version of `getOutputParameterIndex` around for compatibility?

**Comment by jbj** [comment]
Tests failed:

```
--- expected
+++ actual
@@ -1,3 +1,4 @@
+WARNING: Predicate variadicFormatter has been deprecated and may be removed in future (variadicFormatter.ql:4,7-24)
 | __vswprintf_l | 1 |
 | _vsnprintf_l | 2 |
 | _vsnprintf_s | 3 |
[6/8] [88/522 comp 2.7s eval 36ms] FAILED(RESULT) /home/runner/work/semmle-code/semmle-code/semmlecode-cpp-tests/library-tests/printf/formattingFunction/variadicFormatter.ql
```

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll]: > Do we need to keep a deprecated parameter-less version of getOutputParameterIndex around for compatibility?

[There is one](https://github.com/github/codeql/pull/4754/files#diff-c81f331ad8237d9b6c277111ba00e160051d9c4ebdb36549cf27314bbe9b99f5R124), it doesn't really stand out in the diff.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll]: Ah, sorry. You're right!

**Comment by geoffw0** [comment]
Created https://github.com/github/semmle-code/pull/38334 (internal link) to fix the failing test.  Both PRs should be merged at about the same time.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Comment by Arsalsaeed** [comment]
C+

**Comment by Arsalsaeed** [comment]
C++

**Comment by Arsalsaeed** [comment]
C++

**Comment by Arsalsaeed** [comment]
C+

**Comment by Arsalsaeed** [comment]
C++

**Pull request merged**